lightning-invoice/src/payment.rs

```diff
@@ -0,0 +1,572 @@
+	router: R,
+	network_graph: G,
+	invoice_cache: Mutex<HashMap<PaymentHash, Invoice>>,
+}
+
+/// A trait defining behavior of an [`Invoice`] payer.
+pub trait Payer {
+	/// Returns the payer's node id.
+	fn node_id(&self) -> PublicKey;
+
+	/// Returns the payer's channels.
+	fn first_hops(&self) -> Vec<ChannelDetails>;
+
+	/// Sends a payment over the Lightning Network using the given [`Route`].
+	fn send_payment(
+		&self, route: &Route, payment_hash: PaymentHash, payment_secret: &Option<PaymentSecret>
+	) -> Result<(), PaymentSendFailure>;
+}
+
+/// A trait defining behavior for routing an [`Invoice`] payment.
+pub trait Router {
+	/// Finds a [`Route`] between `payer` and `payee` using the provided first and last hops and
+	/// [`NetworkGraph`] for a payment with the given values.
+	fn find_route(
+		&self,
+		payer: &PublicKey,
+		network: &NetworkGraph,
+		payee: &PublicKey,
+		payee_features: Option<InvoiceFeatures>,
+		first_hops: Option<&[&ChannelDetails]>,
+		last_hops: &[&RouteHint],
+		final_value_msat: u64,
+		final_cltv: u32,
+	) -> Result<Route, LightningError>;
+}
+
+/// An error that may occur when making a payment.
+#[derive(Debug)]
+pub enum PaymentError {
+	/// An error resulting from the provided [`Invoice`] or payment hash.
+	Invoice(&'static str),
+	/// An error occurring when finding a route.
+	Routing(LightningError),
+	/// An error occurring when sending a payment.
+	Sending(PaymentSendFailure),
+}
+
+/// An [`EventHandler`] decorator for retrying failed payments.
+pub struct PaymentRetryHandler<I, P: Deref, R, G, L: Deref, E>
+where
+	I: Deref<Target = InvoicePayer<P, R, G>>,
+	P::Target: Payer,
+	R: Router,
+	G: Deref<Target = NetworkGraph>,
+	L::Target: Logger,
+	E: EventHandler,
+{
+	invoice_payer: I,
+	payment_attempts: RefCell<HashMap<PaymentHash, usize>>,
+	logger: L,
+	event_handler: E,
+}
+
+impl<P: Deref, R, G> InvoicePayer<P, R, G>
+where P::Target: Payer, R: Router, G: Deref<Target = NetworkGraph> {
+	/// Creates an invoice payer.
+	pub fn new(payer: P, router: R, network_graph: G) -> Self {
+		Self {
+			payer,
+			router,
+			network_graph,
+			invoice_cache: Mutex::new(HashMap::new()),
+		}
+	}
+
+	/// Pays the given [`Invoice`], caching it for later use if a retry is needed.
+	pub fn pay_invoice(&self, invoice: &Invoice) -> Result<(), PaymentError> {
+		let payment_hash = PaymentHash(invoice.payment_hash().clone().into_inner());
+		let mut invoice_cache = self.invoice_cache.lock().unwrap();
+		match invoice_cache.entry(payment_hash) {
+			hash_map::Entry::Vacant(entry) => {
+				self.pay_invoice_internal(invoice)?;
+				entry.insert(invoice.clone());
+				Ok(())
+			},
+			hash_map::Entry::Occupied(_) => Err(PaymentError::Invoice("payment pending")),
+		}
+	}
+
+	fn pay_cached_invoice(&self, payment_hash: &PaymentHash) -> Result<(), PaymentError> {
+		let invoice_cache = self.invoice_cache.lock().unwrap();
+		match invoice_cache.get(payment_hash) {
+			Some(invoice) => self.pay_invoice_internal(invoice),
```

Comment:
Should we somehow check that the route used on a retry is different from the original? I could see users using this API and just hooking it up to a remote route-getter which doesn't care about our failed payments.

Reply:
I was hoping that for simplicity this could be route-agnostic. Any changes to the network or scoring could affect the route calculation, so we should let those modules (e.g., higher-level event handlers) take care of making adjustments for the next time a route is fetched, if necessary, even if it means the same route is tried.